specs/core/0_beacon-chain.md

```diff
@@ -1739,61 +1691,87 @@ def update_validator_registry(state: BeaconState) -> None:
     # Activate validators within the allowable balance churn
     balance_churn = 0
     for index, validator in enumerate(state.validator_registry):
-        if validator.status == PENDING_ACTIVATION and state.validator_balances[index] >= MAX_DEPOSIT * GWEI_PER_ETH:
+        if validator.activation_slot > state.slot and state.validator_balances[index] >= MAX_DEPOSIT * GWEI_PER_ETH:
```

Comment:
This should be `if validator.activation_slot == FAR_FUTURE_SLOT`. Because `activation_slot` is set for a future slot, this loop could re-activate already pending active validators in a future epoch transition (if a registry changed occurred within `ENTRY_EXIT_SLOTS` from the current transition)

Reply:
@vbuterin This is actually a problem. Should be `validator.activation_slot > state.slot + ENTRY_EXIT_DELAY`

Look at the following scenario with `validator`:
* `state.slot` is 256. A validator registry change occurs
    * `validator` is not activated so `activation_slot == FAR_FUTURE_SLOT`
    * in this loop, `validator` has `activation_slot > stat.slot` and gets activated
    * `validator.activation_slot == 512` (`256 + ENTRY_EXIT_DELAY`)
*  Next epoch -- `state.slot` is now 320. A registry change occurs
    * `validator.activation_slot == 512`
    * in this loop, `validator.activation_slot > state.slot` (`512 > 320`) so gets activated again
    * `validator.activation_slot == 576`

Reply:
Yes, you're right. Will fix.